### PR TITLE
[Merged by Bors] - feat: port Data.Set.Ncard

### PR DESCRIPTION
**Pull request by apnelson1**
A port of the newly added `ncard` file. 

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Set/Ncard.lean [hunk 1]
```diff
@@ -0,0 +1,686 @@
+/-
+Copyright (c) 2023 Peter Nelson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Peter Nelson
+
+! This file was ported from Lean 3 source module data.set.ncard
```
Comment (jcommelin) [inline Mathlib/Data/Set/Ncard.lean:6]:
```suggestion
! This file was ported from Lean 3 source module data.set.ncard
! leanprover-community/mathlib commit 74c2af38a828107941029b03839882c5c6f87a04
! Please do not edit these lines, except to modify the commit id
! if you have ported upstream changes.
```

**Comment by jcommelin** [comment]
@apnelson1 I reverted a change that you made to the porting header at the top of the file. This might create a tiny race condition with local commits that you have, but a `git pull` should be able to handle it.

This header needs to follow a strict format, otherwise some of our little robot friends start crying.

File: Mathlib/Data/Set/Ncard.lean [hunk 2]
```diff
@@ -0,0 +1,674 @@
...
+lemma diff_nonempty_of_ncard_lt_ncard (h : s.ncard < t.ncard) (hs : s.Finite := by toFinite_tac) :
+    (t \ s).Nonempty := by
+  rw [Set.nonempty_iff_ne_empty, Ne.def, diff_eq_empty]
+  exact λ h' ↦ h.not_le (ncard_le_of_subset h' hs)
+
+lemma exists_mem_not_mem_of_ncard_lt_ncard (h : s.ncard < t.ncard)
+  (hs : s.Finite := by toFinite_tac) : ∃ e, e ∈ t ∧ e ∉ s :=
+  diff_nonempty_of_ncard_lt_ncard h hs
+
+@[simp] lemma ncard_inter_add_ncard_diff_eq_ncard (s t : Set α) (hs : s.Finite := by toFinite_tac) :
```
Comment (ChrisHughes24) [inline Mathlib/Data/Set/Ncard.lean:474]:
```suggestion
@[simp]
lemma ncard_inter_add_ncard_diff_eq_ncard (s t : Set α) (hs : s.Finite := by toFinite_tac) :
```

File: Mathlib/Data/Set/Ncard.lean [hunk 3]
```diff
@@ -0,0 +1,674 @@
...
+  · rw [ncard_insert_of_mem h, if_pos h]
+  · rw [ncard_insert_of_not_mem h hs, if_neg h]
+
+lemma ncard_le_ncard_insert (a : α) (s : Set α) : s.ncard ≤ (insert a s).ncard := by
+  classical
+  refine'
+    s.finite_or_infinite.elim (λ h ↦ _) (λ h ↦ by (rw [h.ncard]; exact Nat.zero_le _))
+  rw [ncard_insert_eq_ite h]; split_ifs <;> simp
+
+@[simp] lemma card_doubleton (h : a ≠ b) : ({a, b} : Set α).ncard = 2 := by
```
Comment (ChrisHughes24) [inline Mathlib/Data/Set/Ncard.lean:158]:
```suggestion
@[simp]
lemma card_doubleton (h : a ≠ b) : ({a, b} : Set α).ncard = 2 := by
```

File: Mathlib/Data/Set/Ncard.lean [hunk 4]
```diff
@@ -0,0 +1,674 @@
...
+  rw [ncard_insert_of_not_mem, ncard_singleton]; simpa
+
+@[simp] lemma ncard_diff_singleton_add_one (h : a ∈ s) (hs : s.Finite := by toFinite_tac) :
```
Comment (ChrisHughes24) [inline Mathlib/Data/Set/Ncard.lean:161]:
```suggestion
@[simp]
lemma ncard_diff_singleton_add_one (h : a ∈ s) (hs : s.Finite := by toFinite_tac) :
```

File: Mathlib/Data/Set/Ncard.lean [hunk 5]
```diff
@@ -0,0 +1,674 @@
...
+
+lemma ncard_singleton_inter : ({a} ∩ s).ncard ≤ 1 := by
+  classical
+  rw [← inter_self {a}, inter_assoc, ncard_eq_toFinset_card _, Finite.toFinset_inter,
+    Finite.toFinset_singleton]
+  · apply Finset.card_singleton_inter
+  all_goals toFinite_tac
+section InsertErase
+
+@[simp] lemma ncard_insert_of_not_mem (h : a ∉ s) (hs : s.Finite := by toFinite_tac) :
```
Comment (ChrisHughes24) [inline Mathlib/Data/Set/Ncard.lean:128]:
```suggestion
@[simp]
lemma ncard_insert_of_not_mem (h : a ∉ s) (hs : s.Finite := by toFinite_tac) :
```

File: Mathlib/Data/Set/Ncard.lean [hunk 6]
```diff
@@ -0,0 +1,674 @@
...
+lemma Finite_of_ncard_ne_zero (hs : s.ncard ≠ 0) : s.Finite :=
+  s.finite_or_infinite.elim id λ h ↦ (hs h.ncard).elim
+
+lemma Finite_of_ncard_pos (hs : 0 < s.ncard) : s.Finite :=
+  Finite_of_ncard_ne_zero hs.ne.symm
+
+lemma nonempty_of_ncard_ne_zero (hs : s.ncard ≠ 0) : s.Nonempty := by
+  rw [nonempty_iff_ne_empty]; rintro rfl; simp at hs
+
+@[simp] lemma ncard_singleton (a : α) : ({a} : Set α).ncard = 1 := by
```
Comment (ChrisHughes24) [inline Mathlib/Data/Set/Ncard.lean:117]:
```suggestion
@[simp]
lemma ncard_singleton (a : α) : ({a} : Set α).ncard = 1 := by
```

File: Mathlib/Data/Set/Ncard.lean [hunk 7]
```diff
@@ -0,0 +1,674 @@
...
+  @Finite.card_le_of_embedding _ _ (finite_coe_iff.mpr ht) (Set.embeddingOfSubset _ _ hst)
+
+lemma ncard_mono [Finite α] : @Monotone (Set α) _ _ _ ncard := λ _ _ ↦ ncard_le_of_subset
+
+@[simp] lemma ncard_eq_zero (hs : s.Finite := by toFinite_tac) :
+    s.ncard = 0 ↔ s = ∅ := by
+  rw [ncard_def, @Finite.card_eq_zero_iff _ hs.to_subtype, isEmpty_subtype,
+    eq_empty_iff_forall_not_mem]
+
+@[simp] lemma ncard_coe_Finset (s : Finset α) : (s : Set α).ncard = s.card := by
```
Comment (ChrisHughes24) [inline Mathlib/Data/Set/Ncard.lean:86]:
```suggestion
@[simp]
lemma ncard_coe_Finset (s : Finset α) : (s : Set α).ncard = s.card := by
```

File: Mathlib/Data/Set/Ncard.lean [hunk 8]
```diff
@@ -0,0 +1,674 @@
...
+    s.ncard = s.toFinset.card := by
+  simp [ncard_def, Nat.card_eq_fintype_card]
+
+lemma ncard_le_of_subset (hst : s ⊆ t) (ht : t.Finite := by toFinite_tac) :
+    s.ncard ≤ t.ncard :=
+  @Finite.card_le_of_embedding _ _ (finite_coe_iff.mpr ht) (Set.embeddingOfSubset _ _ hst)
+
+lemma ncard_mono [Finite α] : @Monotone (Set α) _ _ _ ncard := λ _ _ ↦ ncard_le_of_subset
+
+@[simp] lemma ncard_eq_zero (hs : s.Finite := by toFinite_tac) :
```
Comment (ChrisHughes24) [inline Mathlib/Data/Set/Ncard.lean:81]:
```suggestion
@[simp]
lemma ncard_eq_zero (hs : s.Finite := by toFinite_tac) :
```

**Review by ChrisHughes24 — CHANGES_REQUESTED** [review]
All the `#align`s have been deleted and ought to be restored.

We haven't been changing `theorem` to `lemma` when porting and we tend to keep the `fun` that the mathport output gives without changing to `λ`, but I don't think we have a rule not to do this either.

**Comment by digama0** [comment]
I would say that `fun` is preferred over `λ`, but `@[simp]` on the same line is okay.

File: Mathlib/Data/Set/Ncard.lean [hunk 9]
```diff
@@ -0,0 +1,773 @@
...
+/-- A tactic (for use in default params) that applies `Set.toFinite` to synthesize a
+  `Set.finite` term. -/
+syntax "toFinite_tac" : tactic
+
+macro_rules
+  | `(tactic| toFinite_tac) => `(tactic| apply Set.toFinite)
+
+namespace Set
+
+/-- The cardinality of `s : set α` . Has the junk value `0` if `s` is Infinite -/
```
Comment (semorrison) [inline Mathlib/Data/Set/Ncard.lean:61]:
```suggestion
/-- The cardinality of `s : Set α` . Has the junk value `0` if `s` is Infinite -/
```

**Review by kim-em — APPROVED** [review]

**Comment by semorrison** [comment]
bors d+

**Comment by Parcly-Taxel** [comment]
(https://github.com/leanprover-community/mathlib4/actions/runs/5189315935/jobs/9354146098)

**Pull request closed without merging**